src/MacroExpander.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * This file contains the “gullet” where macros are expanded
+ * until only non-macro tokens remain.
+ */
+
+var Lexer = require("./Lexer");
+
+function MacroExpander(input, macros) {
+    this.lexer = new Lexer(input);
+    this.macros = macros || {};
+    this.stack = []; // contains tokens in REVERSE order
+    this.discardedWhiteSpace = [];
+}
+
+MacroExpander.prototype.nextToken = function() {
+    for (;;) {
+        if (this.stack.length === 0) {
+            this.stack.push(this.lexer.lex());
+        }
+        var top = this.stack.pop();
+        var name = top.text;
+        if (!(name.charAt(0) === "\\" && this.macros.hasOwnProperty(name))) {
+            return top;
+        }
+        var expansion = this.macros[name];
+        if (typeof expansion === "string") {
+            var bodyLexer = new Lexer(expansion);
+            expansion = [];
+            var tok = bodyLexer.lex();
+            while (tok.text !== "EOF") {
+                expansion.push(tok);
+                tok = bodyLexer.lex();
+            }
+            expansion.reverse();
```

Comment:
Well, the whole set of current tokens is a stack, and while the `push` and `pop` methods operate on the tail of the JavaScript array, they operate on the head of the TeX token sequence. I guess that `unshift` and `shift` are more costly in most implementations, thus reversing macro bodies when they first get lexed seemed cheaper, performance-wise. Would the following be an appropriate comment to sum this up?

```js
expansion.reverse(); // to fit in with tail-modifying stack
```